transit.md

```diff
@@ -217,3 +272,108 @@ both sides have their own nonce.
+Then, both sides try to connect to each of the peer's IPs. For every connection attempt, they bind
+a new socket on the same port again. Also you read that right, there are no `listen` or `accept`
+calls involved.
+
+This works only with firewalls that silently filter packets and which don't send back any RST packets.
+Unfortunately, this is the default behavior of the kernel when no firewall is active (or the firewall allowed)
+the packets. This means that the firewall hole punching method does *not* work when *no* firewall is
+active (Fun fact: if one side has a firewall and the other side doesn't then connection attempts will
+work or fail depending on which side sent their packets first—I'm really glad for you that you didn't
+have to debug this …). Thus, **this method can only be an addition to the "easy" way** specified above.
+Sadly, this means that the number of hints – and thus, attempted connections – is doubled. If you manage
+to listen and accept on the same port so that only one set of hints needs to be sent out, please contact
+`piegames` with your findings. Also read [this paper](https://bford.info/pub/net/p2pnat/) (Python code
+[here](https://github.com/dwoz/python-nat-hole-punching/)) for more details on the subject.
+
+### Traversing NAT
+
+To additionally traverse through a potential NAT, clients need to query their external IP address
+from a STUN server\*. Because of how NATs work, the socket must again be bound with `SO_REUSEADDR`
+and its port must be used for the hint. Clients may `shutdown` the connection to the STUN server, but
+the socket must be kept open. No enhanced NAT detection or other advanced STUN features need to be
+used: we only care about the external IP address and if it fails we can always simply fall back to
+the relay. Generally, a failure to do STUN should be silent and the query should time out rather
+quickly.
+
+Recall that NATs are a structure of the legacy and that IPv6 users usually have a globally routed
+public address, for which simply firewall hole punching is sufficient. For this reason, you only need
+to do an IPv4 query and you always want to prefer IPv6 connections.
+
+\* Since you should always host your own for production use (and also most of the public ones don't
+support TCP), we provide <tcp://stun.magic-wormhole.io:3478> as part of the Magic Wormhole infrastructure
+(please only use for Magic Wormhole clients).
+
+### Choosing the best connection
+
+When multiple connection attempts go through, it is up to the leader side to select the one to use and
+to cancel the other ones. A naive implementation might simply use the first one that got established,
+in the assumption that it will be the best (it probably had the best round trip time, modulo some jitter).
+But next to that, there are other criteria that might be taken in on the rating:
+
+1. Prefer direct connections over relay ones
+2. Prefer IPv6 over IPv4
```

Comment:
For 1 and 3 the justification seems intuitive. It's less clear why v4 vs v6 matters? (At least, I feel this should be "3" since preferring local connections seems more important)